tools/tpm2_policycountertimer.c

```diff
@@ -0,0 +1,251 @@
+
+    TPM2B_OPERAND operand_b;
+
+    uint16_t offset;
+
+    TPM2_EO operation;
+
+    //Output
+    const char *policy_digest_path;
+};
+
+static tpm_policycountertimer_ctx ctx = {
+    .operation = TPM2_EO_EQ,
+    .offset = 0,
+};
+
+static bool process_time_info(char *value, uint16_t offset, uint8_t size) {
+
+    ctx.offset = offset;
+
+    uint64_t datatype64 = 0;
+    uint32_t datatype32 = 0;
+    bool result = true;
+
+    char *string_bytes = malloc( 1 + (size * 2) );
+    if (!string_bytes) {
+        LOG_ERR("oom");
+        return false;
+    }
+
+    switch(size) {
+        case sizeof(uint64_t):
+            result = tpm2_util_string_to_uint64(value, &datatype64);
+            if (!result) {
+                LOG_ERR("Couldn't convert starting offset, got: \"%s\"", value);
+                goto process_time_info_out;
+            }
+            snprintf(string_bytes, (1 + (size * 2)), "%016lX", datatype64);
+            break;
+        case sizeof(uint32_t):
+            result = tpm2_util_string_to_uint32(value, &datatype32);
+            if (!result) {
+                LOG_ERR("Couldn't convert starting offset, got: \"%s\"", value);
+                goto process_time_info_out;
+            }
+            snprintf(string_bytes, (1 + (size * 2)), "%08X", datatype32);
+            break;
+    }
+
+    ctx.operand_b.size = size;
+    int rc = tpm2_util_hex_to_byte_structure(string_bytes, &ctx.operand_b.size,
+        ctx.operand_b.buffer);
+    if (rc) {
+        LOG_ERR("Error converting the milliseconds input.");
+        result = false;
+    }
+
+process_time_info_out:
+    free(string_bytes);
+    return result;
+}
+
+#define inline_process_time_info(argc, argv, arg_ctr, offset, datasize) \
+    if (argc == (arg_ctr + 1)) { \
+        LOG_ERR("Must specify a value."); \
+        return false; \
+    } \
+    bool result = process_time_info(argv[arg_ctr + 1], offset, datasize); \
+    if (!result) { \
+        return false; \
+    }
+
+static bool on_arg(int argc, char **argv) {
+
+    if (argc > 2) {
+        LOG_ERR("Invalid input. Refer the man page.");
+        return false;
+    }
+
+    uint8_t arg_ctr = 0;
+    do {
+        if (!strcmp("time", argv[arg_ctr])) {
+            inline_process_time_info(argc, argv, arg_ctr, 0, sizeof(uint64_t));
+            arg_ctr++;
+        } else if (!strcmp("clock", argv[arg_ctr])) {
+            inline_process_time_info(argc, argv, arg_ctr, 8, sizeof(uint64_t));
+            arg_ctr++;
+        } else if (!strcmp("resets", argv[arg_ctr])) {
+            inline_process_time_info(argc, argv, arg_ctr, 16, sizeof(uint32_t));
+            arg_ctr++;
+        } else if (!strcmp("restarts", argv[arg_ctr])) {
+            inline_process_time_info(argc, argv, arg_ctr, 20, sizeof(uint32_t));
+            arg_ctr++;
+        } else if (!strcmp("safe", argv[arg_ctr])) {
+            ctx.offset = 24;
+            ctx.operand_b.size = 1;
+            ctx.operand_b.buffer[0] = TPM2_YES;
```

Comment:
why? Explain this magic in a comment

Reply:
essentially these magic numbers are the offsets in the TPMS_TIME_INFO structure. Although I must have a macro definition for the offset values, I agree.

Reply:
Populate a tpms_time_info struct and the use libmu to convert it to a byte array

Reply:
TPM expects only a single structure-member data in the operand buffer.